### PR TITLE
Delete method for SubjectController

### PR DESCRIPTION
**Pull request by AnyaMaz**
#365 
- [X] Add Delete method
- [X] Add tests for method

### CODE REVIEW

**Comment by AnyaMaz** [comment]
@malinink Check this PR please.

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 1]
```diff
@@ -66,7 +68,26 @@
...
+    @DeleteMapping("/{subjectId}")
+    public void delete(@PathVariable Integer subjectId) throws NotFoundException, ForbiddenException {
+        final SubjectModelInterface subjectModel = this.getSubjectModel(subjectId);
+        if (subjectModel == null) {
+            throw new NotFoundException();
+        }
+        if (!this.acl.hasAccess(subjectModel)) {
+            throw new ForbiddenException();
+        }
+        this.subjectsService.delete(new SubjectsOptions().withRelations(subjectModel));
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: Why do you use SubjectOptions like that? There is no need.

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 2]
```diff
@@ -66,7 +68,26 @@
...
+    }
+
+    private SubjectModelInterface getSubjectModel(Integer id, SubjectsOptionsInterface subjectOpt) {
+        final SubjectModelInterface subjectModel = this.subjectsService.find(id, subjectOpt);
+        return subjectModel;
+    }
+
+    private SubjectModelInterface getSubjectModel(Integer id) throws NotFoundException {
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: I don't think that getSubjectModel mathod is good for fething it for delete.
Use getSubject as always - for Auth.

And if you want to fetch subject for delete manually pass this.subjectsOptionsBuilder.forDelete() into ` getSubjectModel(Integer id, SubjectsOptionsInterface subjectOpt)` in delete method.

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 3]
```diff
@@ -66,7 +68,26 @@
...
+        if (subjectModel == null) {
+            throw new NotFoundException();
+        }
+        if (!this.acl.hasAccess(subjectModel)) {
+            throw new ForbiddenException();
+        }
+        this.subjectsService.delete(new SubjectsOptions().withRelations(subjectModel));
+    }
+
+    private SubjectModelInterface getSubjectModel(Integer id, SubjectsOptionsInterface subjectOpt) {
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: `subjectOpt` is not good variable name

File: src/test/java/easytests/api/v1/controllers/SubjectsControllerTest.java [hunk 4]
```diff
@@ -139,8 +141,38 @@ public void testListForbidden() throws Exception {
...
      * show(subjectId)
      */
-    /**
-     * delete(subjectId)
-     */
-    
+    @Test
+    public void testDeleteSuccess() throws Exception {
+        final SubjectModelInterface subjectModel = this.subjectsSupport.getModelFixtureMock(0);
+        when(this.subjectsService.find(any(), any())).thenReturn(subjectModel);
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/SubjectsControllerTest.java]: also we need to be ensure that object with specified ID passed into delete method of service.

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by AnyaMaz** [comment]
@malinink check pr plz

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 5]
```diff
@@ -93,4 +93,16 @@ private SubjectModelInterface getSubjectModel(Integer id) throws NotFoundExcepti
     /**
      * delete(subjectId)
      */
+
+    @DeleteMapping("/{subjectId}")
+    public void delete(@PathVariable Integer subjectId) throws NotFoundException, ForbiddenException {
+        final SubjectModelInterface subjectModel = this.getSubjectModel(subjectId);
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: get subject model will not fetch it for delete, you need to provide Options to select object for delete.
Select for delete (All topics, questions and etc will be included and linked)
Than, you need call delete with that option.


So:
You need first to get subjects for auth, verify user have access and subjects exists.
Than select subject for delete.
Than delete subject with for delete options.

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by AnyaMaz** [comment]
@malinink Rdy For Review

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 6]
```diff
@@ -90,7 +90,22 @@ private SubjectModelInterface getSubjectModel(Integer id, SubjectsOptionsInterfa
...
+        return this.getSubjectModel(id, this.subjectsOptionsBuilder.forDelete());
+    }
     /**
      * delete(subjectId)
      */
+
+    @DeleteMapping("/{subjectId}")
+    public void delete(@PathVariable Integer subjectId) throws NotFoundException, ForbiddenException {
+        final SubjectModelInterface subjectModelforAuth = this.getSubjectModel(subjectId);
+        System.err.println("TRALALA");
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: remove

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 7]
```diff
@@ -90,7 +90,22 @@ private SubjectModelInterface getSubjectModel(Integer id, SubjectsOptionsInterfa
...
+        if (!this.acl.hasAccess(subjectModelforAuth)) {
+            throw new ForbiddenException();
+        }
+        final SubjectModelInterface subjectModelforDelete = this.getSubjectModelforDelete(subjectId);
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: использовать отдельную переменную для опшенов, исподзовать её для получения и удаления

File: src/test/java/easytests/api/v1/controllers/SubjectsControllerTest.java [hunk 8]
```diff
@@ -189,5 +191,41 @@ public void testShowForbidden() throws Exception {
     /**
      * delete(subjectId)
      */
-    
+    @Test
+    public void testDeleteSuccess() throws Exception {
+        final SubjectModelInterface subjectModel = this.subjectsSupport.getModelFixtureMock(0);
+        when(this.subjectsService.find(any(), any())).thenReturn(subjectModel);
+        when(this.acl.hasAccess(any(SubjectModelInterface.class))).thenReturn(true);
+        final ArgumentCaptor<SubjectModelInterface> argumentCaptor = ArgumentCaptor.forClass(SubjectModelInterface.class);
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/SubjectsControllerTest.java]: не нужен

File: src/test/java/easytests/api/v1/controllers/SubjectsControllerTest.java [hunk 9]
```diff
@@ -189,5 +191,41 @@ public void testShowForbidden() throws Exception {
...
+        this.mvc.perform(delete("/v1/subjects/1")
+                .contentType(MediaType.APPLICATION_JSON)
+        )
+                .andExpect(status().isForbidden())
+                .andExpect(content().string(""))
+                .andReturn();
+    }
+    @Test
+    public void testDeleteNotFound() throws Exception {
+        when(this.acl.hasAccess(any(SubjectModelInterface.class))).thenReturn(true);
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/SubjectsControllerTest.java]: мок на null

**Review by malinink — CHANGES_REQUESTED** [review]

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 10]
```diff
@@ -70,7 +70,7 @@
 
     @GetMapping("/{subjectId}")
     public Subject show(@PathVariable Integer subjectId) throws NotFoundException, ForbiddenException {
-        final SubjectModelInterface subjectModel = this.getSubjectModel(subjectId);
+        final SubjectModelInterface subjectModel = this.getSubjectModel(subjectId, subjectsOptionsBuilder.forAuth());
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: нет необходимости, `this.getSubjectModel(subjectId);` и так выберет forAuth. получается лаконичнее.

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 11]
```diff
@@ -87,10 +87,19 @@ private SubjectModelInterface getSubjectModel(Integer id, SubjectsOptionsInterfa
         return subjectModel;
     }
 
-    private SubjectModelInterface getSubjectModel(Integer id) throws NotFoundException {
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: оставьте метод, он пригодится как минимум в update.

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 12]
```diff
@@ -87,10 +87,19 @@ private SubjectModelInterface getSubjectModel(Integer id, SubjectsOptionsInterfa
...
-        return this.getSubjectModel(id, this.subjectsOptionsBuilder.forAuth());
-    }
     /**
      * delete(subjectId)
      */
+
+    @DeleteMapping("/{subjectId}")
+    public void delete(@PathVariable Integer subjectId) throws NotFoundException, ForbiddenException {
+        final SubjectModelInterface subjectModelforAuth
+                = this.getSubjectModel(subjectId, subjectsOptionsBuilder.forAuth());
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: да и тут его используйте!

File: src/main/java/easytests/api/v1/controllers/SubjectsController.java [hunk 13]
```diff
@@ -87,10 +87,19 @@ private SubjectModelInterface getSubjectModel(Integer id, SubjectsOptionsInterfa
...
+        if (!this.acl.hasAccess(subjectModelforAuth)) {
+            throw new ForbiddenException();
+        }
+        final SubjectModelInterface subjectModelforDelete
+                = this.getSubjectModel(subjectId, subjectsOptionsBuilder.forDelete());
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/SubjectsController.java]: создайте лучге отдельную переменную под SubjectOptions

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by AnyaMaz** [comment]
@malinink ready for review

File: src/test/java/easytests/api/v1/controllers/SubjectsControllerTest.java [hunk 14]
```diff
@@ -189,5 +192,42 @@ public void testShowForbidden() throws Exception {
...
+        subjectModel.map(subjectEntity);
+        when(this.subjectsService.find(any(), any())).thenReturn(subjectModel);
+        when(this.acl.hasAccess(any(SubjectModelInterface.class))).thenReturn(true);
+        this.mvc.perform(delete("/v1/subjects/1")
+                .contentType(MediaType.APPLICATION_JSON)
+        )
+                .andExpect(status().isOk())
+                .andExpect(content().string(""))
+                .andReturn();
+        verify(this.subjectsService, times(1)).delete(subjectModel, subjectsOptionsBuilder.forDelete()  );
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/SubjectsControllerTest.java]: `subjectsOptionsBuilder.forDelete()` -т.к.  subjectsOptionsBuilder это мок - то метод вернет null.
Я бы создал опшен, настроил его на возвращение в методе forDelete, и проверил что вызывается именно с этим объектом метод delete.

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by AnyaMaz** [comment]
@malinink check plz

**Review by malinink — APPROVED** [review]

**Pull request merged**